### PR TITLE
Fetch live sessions based on time table

### PR DESCRIPTION
**Pull request by Bahugunajii**
### Description
- Fetch live sessions according to the time table

### Checklist
- [x] Self-testing

### CODE REVIEW

File: app/page.tsx [hunk 1]
```diff
@@ -53,9 +57,12 @@ export default function Home() {
       const flattenedGroupSessions = groupSessions.flat();
 
       const sessionsData = await Promise.all(flattenedGroupSessions.map(async (groupSession: GroupSession) => {
-        const sessionData = await getSessions(groupSession.session_id);
-        const isActive = sessionData.is_active;
-        const repeatSchedule = sessionData.repeat_schedule;
+        const sessionData = await getSessionSchedule(groupSession.session_id, batchId);
```
Comment (suryabulusu) [inline app/page.tsx:60]: const sessionScheduleData =

File: utils/dateUtils.ts [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
-    const hours = String(date.getUTCHours()).padStart(2, "0");
-    const minutes = String(date.getUTCMinutes()).padStart(2, "0");
+    let date;
+    if (dateTimeStr.includes('AM') || dateTimeStr.includes('PM')) {
+        const [time, period] = dateTimeStr.split(' ');
+        const [hoursStr, minutesStr] = time.split(':');
+        let hours = parseInt(hoursStr);
+        const minutes = parseInt(minutesStr);
+
+        if (period === 'PM' && hours < 12) {
```
Comment (pritamps) [inline utils/dateUtils.ts:9]: Isn't this possible by library methods on datetime like python?

**Review by pritamps — CHANGES_REQUESTED** [review]
Just minor changes!

File: app/page.tsx [hunk 3]
```diff
@@ -75,16 +82,18 @@ export default function Home() {
...
 
     const sessionTime = new Date(`2000-01-01T${sessionTimeStr}`);
+    const sessionEndTime = new Date(`2000-01-01T${sessionEndTimeStr}`);
     const currentTimeObj = new Date(`2000-01-01T${currentTimeStr}`);
     const timeDifference = (sessionTime.getTime() - currentTimeObj.getTime()) / (1000 * 60);
 
-    if (data.platform === 'meet') {
-      if (timeDifference <= 5) {
+    if (data.session && data.session.platform === 'meet') {
+      if (timeDifference <= 5 && sessionEndTime.getTime() > currentTimeObj.getTime()) {
```
Comment (suryabulusu) [inline app/page.tsx:94]: maybe call these variables something else for clear understanding. 

const minutesUntilSessionStart = (sessionStartTime.getTime() - currentTimeObj.getTime()) / (1000 * 60);

const hasSessionNotEnded = sessionEndTime.getTime() > currentTimeObj.getTime()

File: app/page.tsx [hunk 4]
```diff
@@ -75,16 +82,18 @@ export default function Home() {
   function renderButton(data: any) {
     const currentTime = new Date();
     const sessionTimeStr = formatSessionTime(data.start_time);
```
Comment (suryabulusu) [inline app/page.tsx:84]: call it session sessionStartTimeStr

File: app/page.tsx [hunk 5]
```diff
@@ -99,11 +108,20 @@ export default function Home() {
       }
     }
     else if (data.redirectPlatform === 'quiz') {
+      if (timeDifference <= 5 && sessionEndTime.getTime() > currentTimeObj.getTime()) {
```
Comment (suryabulusu) [inline app/page.tsx:111]: same comments as above

File: app/types.ts [hunk 6]
```diff
@@ -140,3 +140,11 @@ export interface QuizSession {
   stream: string,
   id: string
 }
+
+export interface SessionSchedule {
```
Comment (suryabulusu) [inline app/types.ts:144]: this really exactly matches with SessionOccurrence haha

**Review by suryabulusu — COMMENTED** [review]
added comments

File: app/types.ts (hunk 6 above)
Comment (suryabulusu) [reply app/types.ts:144]: shouldn't a batch_id be here somewhere

File: api/afdb/session.ts [hunk 7]
```diff
@@ -65,6 +65,26 @@ export const getSessions = async (sessionId: number) => {
   }
 };
 
+export const getSessionSchedule = async (sessionId: number, batchId?: number) => {
+  try {
+    const queryParams = new URLSearchParams();
+    if (sessionId !== undefined) queryParams.append('session_id', sessionId.toString());
+    if (batchId !== undefined) queryParams.append('batch_id', batchId.toString());
+    const urlWithParams = `${url}/session-schedule?${queryParams.toString()}`;
```
[the remark this answers is not in the document]
Comment (Bahugunajii) [reply api/afdb/session.ts:73]: If they both are undefined, then queryParams just becomes empty. So it won't throw an error.

File: app/page.tsx [hunk 8]
```diff
@@ -65,7 +72,7 @@ export default function Home() {
 
       const filteredSessions = sessionsData.filter(session => session !== null);
 
-      const liveClassesData = filteredSessions.filter((session: Session) => session.platform === 'meet');
+      const liveClassesData = filteredSessions.filter((sessionSchedule: SessionSchedule) => sessionSchedule.session.platform === 'meet');
```
[the remark this answers is not in the document]
Comment (Bahugunajii) [reply app/page.tsx:75]: I think a dynamically fetched data (`sessionSchedule.session.platform`) should be compared with a predefined value here ('meet`). Maybe we can define these constants in a separate file like `config.ts`. We do have a source to get session's platform details in LHS here but we don't have anything for RHS (also we don't have platform details in session's purpose)

File: utils/dateUtils.ts (hunk 2 above)
Comment (Bahugunajii) [reply utils/dateUtils.ts:9]: I don't think so 😐

File: api/afdb/session.ts [hunk 9]
```diff
@@ -65,6 +65,26 @@ export const getSessions = async (sessionId: number) => {
...
+    const response = await fetch(urlWithParams, getFetchConfig(bearerToken));
+
+    if (!response.ok) {
+      throw new Error(`Error in fetching Session Schedule Details: ${response.statusText}`);
+    }
+
+    const data = await response.json();
```
[the remark this answers is not in the document]
Comment (Bahugunajii) [reply api/afdb/session.ts:80]: Without await, the code would continue executing before the JSON data is available, which could lead to errors or incorrect behavior when trying to access the data.

File: app/types.ts (hunk 6 above)
Comment (Bahugunajii) [reply app/types.ts:144]: Added!

File: app/page.tsx (hunk 5 above)
Comment (Bahugunajii) [reply app/page.tsx]: Made suggested changes!

File: app/page.tsx (hunk 3 above)
Comment (Bahugunajii) [reply app/page.tsx]: Made suggested changes!

File: api/afdb/session.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (pritamps) [reply api/afdb/session.ts:73]: But you can just not make the request, then, right?

File: api/afdb/session.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (pritamps) [reply api/afdb/session.ts:80]: Isn't the JSON 100% available after the request is done? Can you check once?

File: app/page.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (pritamps) [reply app/page.tsx:75]: I'm saying don't compare to platform, and instead compare to purpose

File: api/afdb/session.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Bahugunajii) [reply api/afdb/session.ts:80]: No.

File: app/page.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bahugunajii) [reply app/page.tsx:75]: Ohh understood 😅

File: api/afdb/session.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Bahugunajii) [reply api/afdb/session.ts:73]: We don't need to explicitly add a check here in the server component. It's like this function won't get called if the value of these arguments is null in client side.

**Review by suryabulusu — APPROVED** [review]
looks good to me

**Pull request merged**